rust/chains/hyperlane-cosmos/src/interchain_gas.rs

```diff
@@ -0,0 +1,183 @@
+use async_trait::async_trait;
+use base64::engine::general_purpose::STANDARD;
+use base64::Engine;
+use cosmrs::tendermint::abci::EventAttribute;
+use hyperlane_core::{
+    ChainResult, ContractLocator, HyperlaneChain, HyperlaneContract, Indexer,
+    InterchainGasPaymaster, SequenceIndexer, U256,
+};
+use hyperlane_core::{HyperlaneDomain, HyperlaneProvider, InterchainGasPayment, LogMeta, H256};
+use std::ops::RangeInclusive;
+use tracing::info;
+
+use crate::grpc::WasmGrpcProvider;
+use crate::rpc::{CosmosWasmIndexer, WasmIndexer};
+use crate::signers::Signer;
+use crate::{ConnectionConf, CosmosProvider};
+
+/// A reference to a InterchainGasPaymaster contract on some Cosmos chain
+#[derive(Debug)]
+pub struct CosmosInterchainGasPaymaster {
+    _conf: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    _signer: Signer,
+    _provider: Box<WasmGrpcProvider>,
+}
+
+impl HyperlaneContract for CosmosInterchainGasPaymaster {
+    fn address(&self) -> H256 {
+        self.address
+    }
+}
+
+impl HyperlaneChain for CosmosInterchainGasPaymaster {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(CosmosProvider::new(self.domain.clone()))
+    }
+}
+
+impl InterchainGasPaymaster for CosmosInterchainGasPaymaster {}
+
+impl CosmosInterchainGasPaymaster {
+    /// create new Cosmos InterchainGasPaymaster agent
+    pub fn new(conf: ConnectionConf, locator: ContractLocator, signer: Signer) -> Self {
+        let provider = WasmGrpcProvider::new(conf.clone(), locator.clone(), signer.clone());
+
+        Self {
+            _conf: conf,
+            domain: locator.domain.clone(),
+            address: locator.address,
+            _signer: signer,
+            _provider: Box::new(provider),
+        }
+    }
+}
+
+/// A reference to a InterchainGasPaymasterIndexer contract on some Cosmos chain
+#[derive(Debug)]
+pub struct CosmosInterchainGasPaymasterIndexer {
+    indexer: Box<CosmosWasmIndexer>,
+}
+
+impl CosmosInterchainGasPaymasterIndexer {
+    /// create new Cosmos InterchainGasPaymasterIndexer agent
+    pub fn new(conf: ConnectionConf, locator: ContractLocator, event_type: String) -> Self {
+        let indexer: CosmosWasmIndexer = CosmosWasmIndexer::new(conf, locator, event_type.clone());
+
+        Self {
+            indexer: Box::new(indexer),
+        }
+    }
+
+    fn get_parser(&self) -> fn(attrs: Vec<EventAttribute>) -> Option<InterchainGasPayment> {
```

Comment:
I'm not the biggest fan of returning closures in these `get_parser` fns. The name is also a bit misleading since it suggests you're getting a struct / object